File: src/simulation/culture/group/CultureAspectCenter.java
package simulation.culture.group;


import kotlin.Pair;
import kotlin.random.Random;
import simulation.culture.group.cultureaspect.*;
import simulation.culture.group.reason.BetterAspectUseReason;
import simulation.culture.group.reason.ConstructReasonsKt;
import simulation.culture.group.reason.Reason;
import simulation.culture.thinking.meaning.ConstructMemeKt;
import simulation.space.resource.Resource;

import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static shmp.random.RandomCollectionsKt.randomElementWithProbability;
import static shmp.random.RandomProbabilitiesKt.testProbability;
import static simulation.Controller.session;

public class CultureAspectCenter {
    private Group group;
    private MutableCultureAspectPool aspectPool;
    private Set<Resource> aestheticallyPleasingResources = new HashSet<>();

    private Set<Reason> reasonsWithSystems = new HashSet<>();

    CultureAspectCenter(Group group, Set<CultureAspect> cultureAspects) {
        this.group = group;
        this.aspectPool = new MutableCultureAspectPool(cultureAspects);
    }

    public MutableCultureAspectPool getAspectPool() {
        return aspectPool;
    }

    public void addCultureAspect(CultureAspect cultureAspect) {
        if (cultureAspect != null) {
            aspectPool.add(cultureAspect);
        }
        if (cultureAspect instanceof AestheticallyPleasingObject) {
            aestheticallyPleasingResources.add(
                    ((AestheticallyPleasingObject) cultureAspect).getResource()
            );
        }
    }

    void useCultureAspects() {
        aspectPool.getAll().forEach(a -> a.use(group.getCultureCenter()));
    }

    void addRandomCultureAspect(Group group) {
        if (!testProbability(session.cultureAspectBaseProbability, session.random)) {
            return;
        }
        CultureAspect cultureAspect = null;
        switch (session.random.nextInt(5)) {
            case 0: {
                cultureAspect = ConstructCultureAspectKt.createDepictObject(
                        group.getCultureCenter().getAspectCenter().getAspectPool().getMeaningAspects(),
                        ConstructMemeKt.constructAndAddSimpleMeme(
                                group.getCultureCenter().getMemePool(),
                                session.random,
                                0.5
                        ),
                        group,
                        session.random
                );
                break;
            }
            case 2: {
                cultureAspect = ConstructCultureAspectKt.createAestheticallyPleasingObject(
                        group.getCultureCenter().getAspectCenter().getAspectPool().getProducedResources()
                                .stream()
                                .map(Pair::getFirst)
                                .filter(r -> !aestheticallyPleasingResources.contains(r))
                                .max(Comparator.comparingInt(r -> r.getGenome().getBaseDesirability()))
                                .orElse(null),
                        group
                );
                break;
            }
            case 3: {//TODO recursively go in dependencies;
                cultureAspect = ConstructCultureAspectKt.constructRitual(
                        ConstructReasonsKt.constructBetterAspectUseReason(
                                group,
                                group.getCultureCenter().getAspectCenter().getAspectPool().getConverseWrappers(),
                                reasonsWithSystems,
                                session.random,
                                5//TODO meh, java
                        ),
                        group,
                        session.random
                );
                break;
            }
            case 4: {
                cultureAspect = ConstructCultureAspectKt.createTale(
                        group,
                        session.templateBase,
                        session.random
                );
                break;
            }
        }
        addCultureAspect(cultureAspect);
    }

    void mutateCultureAspects() {
        if (!testProbability(session.groupCultureAspectCollapse, session.random)) {
            return;
        }
        switch (session.random.nextInt(3)) {
            case 0: {
                joinSimilarRituals();
                break;
            } case 1: {
                joinSimilarTalesBy("!actor");
//                joinSimilarTalesBy("@verb");//TODO debug off
                break;
            } case 2: {
                addCultureAspect(
                        ChangeCultureAspectsKt.takeOutDeity(aspectPool, group, session.random)
                );
                break;
            }
        }
    }

    void joinSimilarRituals() {
        RitualSystem system = ChangeCultureAspectsKt.takeOutSimilarRituals(aspectPool, group, 3);
        if (system != null) {
            addCultureAspect(system);
            reasonsWithSystems.add(system.getReason());
        }
    }

    void joinSimilarTalesBy(String infoTag) {
        TaleSystem system = ChangeCultureAspectsKt.takeOutSimilarTalesByTag(infoTag, aspectPool, group, 3);
        if (system != null) {
            addCultureAspect(system);
        }
    }

    public List<CultureAspect> getNeighbourCultureAspects() {
        return group.getRelationCenter().getRelatedGroups().stream()
                .flatMap(g -> g.getCultureCenter().getCultureAspectCenter().getAspectPool().getAll().stream())
                .collect(Collectors.toList());
    }

    public List<CultureAspect> getNeighbourCultureAspects(Predicate<CultureAspect> predicate) {
        return getNeighbourCultureAspects().stream()
                .filter(predicate)
                .collect(Collectors.toList());
    }

    void adoptCultureAspects() {
        if (!session.isTime(session.groupTurnsBetweenAdopts)) {
            return;
        }
        List<CultureAspect> cultureAspects = getNeighbourCultureAspects(a -> !aspectPool.contains(a));
        if (!cultureAspects.isEmpty()) {
            CultureAspect aspect = randomElementWithProbability(
                    cultureAspects,
                    a -> group.getRelationCenter().getNormalizedRelation(a.getGroup()),
                    session.random
            );
            addCultureAspect(aspect);
        }
    }
}

File: src/shmp/simulation/culture/group/centers/CultureAspectCenter.kt
package shmp.simulation.culture.group.centers

import shmp.random.SampleSpaceObject
import shmp.random.randomElementOrNull
import shmp.random.singleton.chanceOf
import shmp.random.singleton.chanceOfNot
import shmp.random.singleton.randomElement
import shmp.random.singleton.testProbability
import shmp.simulation.Controller.session
import shmp.simulation.culture.group.centers.util.*
import shmp.simulation.culture.group.cultureaspect.*
import shmp.simulation.culture.group.cultureaspect.reasoning.*
import shmp.simulation.culture.group.cultureaspect.reasoning.convertion.*
import shmp.simulation.culture.group.cultureaspect.worship.Worship
import shmp.simulation.culture.group.reason.Reason
import shmp.simulation.culture.group.reason.constructBetterAspectUseReason
import shmp.simulation.culture.thinking.meaning.MemeSubject
import shmp.simulation.culture.thinking.meaning.constructAndAddSimpleMeme
import shmp.simulation.space.resource.Resource
import java.util.*
import kotlin.math.pow


class CultureAspectCenter(val reasonField: ReasonField, private val reasonConversions: List<ReasonConversion>) {
    val aspectPool = MutableCultureAspectPool(mutableSetOf())
    private val aestheticallyPleasingResources: MutableSet<Resource> = HashSet()
    private val reasonsWithSystems: MutableSet<Reason> = HashSet()


    internal fun update(group: Group) {
        useCultureAspects(group)
        addRandomCultureAspect(group)
        mutateCultureAspects(group)
        updateReasonings(group)
    }

    private fun updateReasonings(group: Group) {
        session.reasoningUpdate.chanceOf {
            val newReasonings = reasonField.update(listOf(MemeSubject(group.name)), reasonConversions)
            processReasonings(newReasonings)
        }
    }

    private fun processReasonings(reasonings: List<Reasoning>) {
        reasonings.forEach { addCultureAspect(it.toConcept()) }
        reasonings.forEach { addCultureAspect(it.toCherishedResource()) }
    }

    fun addCultureAspect(cultureAspect: CultureAspect?) {
        cultureAspect ?: return

        aspectPool.add(cultureAspect)

        if (cultureAspect is CherishedResource)
            aestheticallyPleasingResources.add(cultureAspect.resource)
    }

    private fun useCultureAspects(group: Group) = aspectPool.all.forEach { it.use(group) }

    fun addRandomCultureAspect(group: Group) {
        session.cultureAspectBaseProbability.chanceOfNot {
            return
        }

        val cultureAspect = when (AspectRandom.values().randomElement()) {
            AspectRandom.Depict -> createDepictObject(
                    group.cultureCenter.aspectCenter.aspectPool.getMeaningAspects(),
                    constructAndAddSimpleMeme(
                            group.cultureCenter.memePool,
                            session.random
                    ),
                    group,
                    session.random
            )
            AspectRandom.AestheticallyPleasing -> createAestheticallyPleasingObject(
                    group.cultureCenter.aspectCenter.aspectPool.producedResources
                            .filter { it.genome.isDesirable }
                            .filter { !aestheticallyPleasingResources.contains(it) }
                            .maxByOrNull { it.genome.baseDesirability },
                    group,
                    session.random
            )
            AspectRandom.Ritual -> createRitual(//TODO recursively go in dependencies;
                    constructBetterAspectUseReason(
                            group,
                            group.cultureCenter.aspectCenter.aspectPool.converseWrappers.sortedBy { it.name },
                            reasonsWithSystems,
                            session.random
                    ),
                    group,
                    session.random
            )
            AspectRandom.Tale -> createTale(
                    group,
                    session.templateBase,
                    session.random
            )
            AspectRandom.Concept -> createSimpleConcept(
                    group,
                    session.random
            )
        }

        addCultureAspect(cultureAspect)
    }

    fun mutateCultureAspects(group: Group) {
        session.groupCultureAspectCollapse.chanceOfNot {
            return
        }

        when (ChangeRandom.values().randomElement()) {
            ChangeRandom.RitualSystem -> joinSimilarRituals()
            ChangeRandom.TaleSystem -> joinSimilarTalesBy("!actor")
            ChangeRandom.Worship -> addCultureAspect(takeOutWorship(aspectPool, session.random))
            ChangeRandom.God -> makeGod(group)
        }
    }

    private fun joinSimilarRituals() = takeOutSimilarRituals(aspectPool)?.let { system ->
        addCultureAspect(system)
        reasonsWithSystems.add(system.reason)
    }

    private fun makeGod(group: Group) = takeOutGod(aspectPool, group, session.random)?.let { cult ->
        addCultureAspect(cult)
    }

    private fun joinSimilarTalesBy(infoTag: String) = takeOutSimilarTalesBy(infoTag, aspectPool)?.let { system ->
        addCultureAspect(system)
    }

    private fun getNeighbourCultureAspects(group: Group) =
            group.relationCenter.relatedGroups.flatMap { n ->
                n.cultureCenter.cultureAspectCenter.aspectPool.all.map { a -> a to n }
            }

    private fun getNeighbourCultureAspects(group: Group, predicate: (CultureAspect) -> Boolean) =
            getNeighbourCultureAspects(group).filter { (f) -> predicate(f) }

    fun adoptCultureAspects(group: Group) {
        if (!session.isTime(session.groupTurnsBetweenAdopts))
            return

        val aspect = randomElementOrNull(
                getNeighbourCultureAspects(group) { !aspectPool.contains(it) },
                { (_, g) -> group.relationCenter.getNormalizedRelation(g) },
                session.random
        )?.first?.adopt(group)
                ?: return

        if (shouldAdopt(aspect))
            addCultureAspect(aspect)
    }

    private fun shouldAdopt(aspect: CultureAspect): Boolean {
        if (aspect is Worship) {
            val similarGodsAmount = aspectPool.worships.count { w ->
                w.worshipObject.memes.any { it in aspect.worshipObject.memes }
            }
            val probability = 1.0 / (similarGodsAmount + 1)
            return probability.pow(2).testProbability()
        }
        return true
    }

    fun die(group: Group) = aspectPool.all.forEach { it.die(group) }

    override fun toString() = """
        |Culture Aspects:
        |${aspectPool.all.joinToString()}
        |
        |Reasons:
        |$reasonField
    """.trimMargin()
}


private enum class AspectRandom(override val probability: Double) : SampleSpaceObject {
    Tale(3.0),
    Depict(1.0),
    AestheticallyPleasing(1.0),
    Ritual(1.0),
    Concept(1.0)
}

private enum class ChangeRandom(override val probability: Double) : SampleSpaceObject {
    RitualSystem(3.0),
    TaleSystem(3.0),
    Worship(2.0),
    God(1.0),
}


fun baseConversions(
        memoryCenter: MemoryCenter,
        aspectCenter: AspectCenter,
        stratumCenter: StratumCenter
) = listOf(
        MemoryConversion(memoryCenter),
        AspectConversion(aspectCenter),
        StratumConversion(stratumCenter),
        CorrespondingConversion,
        OppositionConversion,
        EqualitySubjectCorrelationConversion,
        ActionConversion.PositiveDriveConversion,
        AspectResourcesConversion,
        StratumResourcesConversion
) + allCorrespondingIdealActionConversions